x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go

```diff
@@ -74,7 +76,13 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
 }
 
 // Dispatch dispatches an action using pre-registered set of handlers.
-func (ad *ActionDispatcher) Dispatch(acker store.FleetAcker, actions ...fleetapi.Action) error {
+func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
+	defer func() {
+		apm.CaptureError(ctx, err).Send()
+		span.End()
+	}()
+
```

Comment:
My suggestion to have a context with the `ad.ctx` cancelation and the span. See [this](https://github.com/elastic/beats/pull/30385/files#r813778664) comment for the reasoning behind this.


```suggestion
// ctx is used here ONLY to carry the span, for cancellation use the cancel
// function of the ActionDispatcher.ctx.
func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
	// Creating a child context that carries both the ad.ctx cancellation and
	// the span from ctx.
	ctx, _ = context.WithCancel(ad.ctx) // ignoring the cancel func as ctx will be cancelled when ever ad.ctx is cancelled.
	ctx = apm.ContextWithSpan(ctx, span)

```